src/authenticationinapp/incrementaldecoder.cpp

```diff
@@ -0,0 +1,57 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "incrementaldecoder.h"
+#include "leakdetector.h"
+
+#include <QTextStream>
+
+constexpr int RADIX = 36;
+
+IncrementalDecoder::IncrementalDecoder(QObject* parent) : QObject(parent) {
+  MVPN_COUNT_CTOR(IncrementalDecoder);
+}
+
+IncrementalDecoder::~IncrementalDecoder() {
+  MVPN_COUNT_DTOR(IncrementalDecoder);
+}
+
+IncrementalDecoder::Result IncrementalDecoder::match(QTextStream& stream,
+                                                     const QString& input) {
+  uint32_t index = 0;
+  QString prev;
+
+  while (!stream.atEnd()) {
```

Comment:
It would be nice if this search could be done a bit more efficiently (eg: binary search) since the password dictionary is sorted and kinda large. Unfortunately the encoding of the password file makes this difficult.